### PR TITLE
[VL] Improving VeloxShuffleWriter performance by just iterating valid partitions

### PR DESCRIPTION
**Pull request by zuochunwei**
## What changes were proposed in this pull request?

The `numPartitions_` is set to `4096` for many jobs in our environment(`Meituan`), and the number of rows in a batch is also `4096`. 

The above configuration will result in some partitions not having rows, we need not iterate for all partitions(4096) but only the used partitions.

In order to avoid unnecessary iteration, I added a member data `partitionUsed` representing the partitions having values and then changed the loop based on all partitions to based on `partitionUsed`.

The performance will not be influenced if the `numPartitions_` configured a small value because the number of valid partitions will never be larger than the number of all partitions.

wait for a test under `meituan`'s environment, I will add the test result later.

(Please fill in changes proposed in this fix)

(Fixes: \#ISSUE-ID)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: cpp/velox/shuffle/VeloxShuffleWriter.cc [hunk 1]
```diff
@@ -623,16 +622,22 @@ arrow::Status VeloxShuffleWriter::buildPartition2Row(uint32_t rowNum) {
...
+
+  // calc valid partition list
+  partitionUsed.clear();
+  for (auto pid = 0; pid != numPartitions_; ++pid) {
+    if (partition2RowCount_[pid] > 0) {
+      partitionUsed.push_back(pid);
+    }
+  }
+
+  std::cout << "[zcw] partitionUsed.size:" << partitionUsed.size() << " numPartitions_:" << numPartitions_ << std::endl;
```
Comment (marin-ma) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:640]: Please remove this log, or change it to debug log if necessary.

Comment (zuochunwei) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc:640]: OK

File: cpp/velox/shuffle/VeloxShuffleWriter.cc [hunk 2]
```diff
@@ -42,7 +42,7 @@ using namespace facebook::velox;
 
 namespace gluten {
 
-#define VELOX_SHUFFLE_WRITER_LOG_FLAG 0
+#define VELOX_SHUFFLE_WRITER_LOG_FLAG 1
```
Comment (marin-ma) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:45]: Please disable this flag.

Comment (zuochunwei) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc:45]: OK

**Comment by zuochunwei** [comment]
@marin-ma ci passed

**Review by marin-ma — APPROVED** [review]

**Comment by marin-ma** [comment]
Thanks! cc: @FelixYBW

**Comment by zuochunwei** [comment]
We made a test for a big job, and the following is the result:

select the longest stage(21) of this job, the time cost(Median) is 7min vs 5.8min

test with this PR
image

test without this PR
image

**Comment by marin-ma** [comment]
> We made a test for a big job, and the following is the result:
> 
> select the longest stage(21) of this job, the time cost(Median) is 7min vs 5.8min
> 
> test with this PR image
> 
> test without this PR image

@zuochunwei Is this the test result for the latest main branch, or an old version of main?

**Pull request merged**